cmd/rpcdaemon/commands/trace_adhoc.go

```diff
@@ -1182,7 +1183,7 @@ func (api *TraceAPIImpl) doCallMany(ctx context.Context, dbtx kv.Tx, msgs []type
 			blockCtx.MaxGasLimit = true
 		}
 		ibs.Reset()
-		blockCtx.L1CostFunc = types.NewL1CostFunc(chainConfig, ibs)
+		blockCtx.L1CostFunc = l1CostFunc
```

Comment:
L1CostFunc is created with params read from the state reader. And in this for loop, states changes are committed to the current DB TX. So I think we can not cache L1CostFunc because it could be changed while loop. What do you think?

Reply:
So `statedb`s lifetime is controversial.

Lets observe the following snippet. You can run here: https://go.dev/play/p/LmiuWJJibgt
```golang
package main

import "fmt"

type LambdaFunc func() uint64

func main() {
	n := uint64(13337)
	counterLambda := createCounterLambda(&n)
	for i := 0; i < 5; i++ {
		res := counterLambda()
		fmt.Println(res)
		n++ 
	}
}

func createCounterLambda(n *uint64) LambdaFunc {
	return func() uint64 {
		(*n)++
		return *n
	}
}
```

Output: 
```
13338
13340
13342
13344
13346
```

The variable `n` in upper snippet corresponds to `statedb`. So we are still using the committed database even if we cache the lambda. So I think caching is ok.

Reply:
oh I thought it read params when it created L1CostFunc, but it was not.. 😅